File: week_2/code/day_8/c_k_closest_points.py
import heapq

class Solution:
    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:
        # Max-heap to store the K closest points. 
        # Since heapq is a min-heap, use negative distance.
        max_heap = []
        
        for x, y in points:
            dist = -(x**2 + y**2)
            if len(max_heap) < k:
                heapq.heappush(max_heap, (dist, x, y))
            elif dist > max_heap[0][0]:
                heapq.heappushpop(max_heap, (dist, x, y))
                
        return [[x, y] for (dist, x, y) in max_heap]

# Example Usage
if __name__ == "__main__":
    sol = Solution()
    print(sol.kClosest([[1, 3], [-2, 2]], 1)) # [[-2, 2]]
    print(sol.kClosest([[3, 3], [5, -1], [-2, 4]], 2)) # [[3, 3], [-2, 4]]

File: week_2/code/day_8/c_k_closest_points.js
const MinHeap = require('./min_heap');

/**
 * @param {number[][]} points
 * @param {number} k
 * @return {number[][]}
 */
var kClosest = function (points, k) {
    // We want a Max-Heap based on distance to keep the K closest.
    // So the comparator should return a > b, simulating Max-Heap.
    // Comparator for MinHeap: (a, b) => a - b
    // Comparator for MaxHeap: (a, b) => b - a

    const maxHeap = new MinHeap((a, b) => {
        const distA = a[0] * a[0] + a[1] * a[1];
        const distB = b[0] * b[0] + b[1] * b[1];
        return distB - distA; // This reverse order turns MinHeap logic into MaxHeap
    });

    for (let point of points) {
        maxHeap.push(point);
        if (maxHeap.size() > k) {
            maxHeap.pop(); // Remove the farthest point
        }
    }

    // Extract all points from heap
    const result = [];
    while (maxHeap.size() > 0) {
        result.push(maxHeap.pop());
    }

    return result;
};

// Example Usage
console.log(kClosest([[1, 3], [-2, 2]], 1)); // [[-2, 2]]
console.log(kClosest([[3, 3], [5, -1], [-2, 4]], 2)); // [[3, 3], [-2, 4]]
